src/V3Param.cpp

```diff
@@ -929,6 +931,113 @@ class ParamProcessor final {
+
+        if (classOrPackageRef) {
+            classp = VN_CAST(classOrPackageRef->classOrPackageSkipp(), Class);
+            if (!classp) return;  // No parameters in packages
+            paramsp = classOrPackageRef->paramsp();
+        } else if (classRefDType) {
+            classp = classRefDType->classp();
+            paramsp = classRefDType->paramsp();
+        } else {
+            nodep->v3fatalSrc("resolveDefaultParams called on node which is not a Class ref");
+        }
+
+        UASSERT_OBJ(classp, nodep, "Class or interface ref has no classp/ifacep");
+
+        // Get the parameter list for this class
+        m_classParams.clear();
+        for (AstNode* stmtp = classp->stmtsp(); stmtp; stmtp = stmtp->nextp()) {
+            if (AstParamTypeDType* paramTypep = VN_CAST(stmtp, ParamTypeDType)) {
+                m_classParams.emplace_back(paramTypep, -1);
+            }
+        }
+
+        // For each parameter, detect either a dependent default (copy from previous param)
+        // or a direct type default (for ex. int). Store dependency index in
+        // m_classParams[i].second, default type in defaultTypeNodes[i].
+        std::vector<AstNodeDType*> defaultTypeNodes(m_classParams.size(), nullptr);
+        for (size_t i = 0; i < m_classParams.size(); ++i) {
+            AstParamTypeDType* const paramTypep = m_classParams[i].first;
+            // Parser places defaults/constraints under childDTypep as AstRequireDType
+            AstRequireDType* const reqDtp = VN_CAST(paramTypep->getChildDTypep(), RequireDType);
+            if (!reqDtp) continue;
+
+            // If default is a reference to another param type, record dependency
+            if (AstRefDType* const refDtp = VN_CAST(reqDtp->subDTypep(), RefDType)) {
+                if (AstParamTypeDType* const sourceParamp
+                    = VN_CAST(refDtp->refDTypep(), ParamTypeDType)) {
+                    for (size_t j = 0; j < m_classParams.size(); ++j) {
+                        if (m_classParams[j].first == sourceParamp) {
+                            m_classParams[i].second = static_cast<int>(j);
+                            break;
+                        }
+                    }
+                    continue;  // dependency handled
+                }
+            }
+
+            // If default is a direct type (for ex. int)
+            // also record dependency
+            if (AstNodeDType* const dtp = VN_CAST(reqDtp->lhsp(), NodeDType)) {
+                defaultTypeNodes[i] = dtp;
+            }
+        }
+
+        // Count existing pins and capture them by index for easy lookup
+        int numPins = 0;
+        std::vector<AstPin*> pinsByIndex(m_classParams.size(), nullptr);
+        {
+            size_t idx = 0;
+            for (AstPin* pinp = paramsp; pinp && idx < m_classParams.size();
+                 pinp = VN_AS(pinp->nextp(), Pin)) {
+                pinsByIndex[idx++] = pinp;
+                numPins++;
+            }
+        }
```

Comment:
```suggestion
        for (AstPin* pinp = paramsp; pinp && idx < m_classParams.size();
             pinp = VN_AS(pinp->nextp(), Pin)) {
            pinsByIndex[numPins++] = pinp;
        }
```

Although perhaps just .push_back the number and numPins can instead use pinsByIndex.size()?